cocotb/build_libs.py

```diff
@@ -0,0 +1,473 @@
+    dist,
+    sim_define,
+    extra_lib=[],
+    extra_lib_dir=[],
+):
+    libvpi = Extension(
+        "libvpi",
+        define_macros=[("VPI_CHECKING", "1")] + [(sim_define, "")],
+        include_dirs=[include_dir],
+        libraries=["gpi", "gpilog"] + extra_lib,
+        library_dirs=[build_dir] + extra_lib_dir,
+        sources=[
+            os.path.join(share_lib_dir, "vpi", "VpiImpl.cpp"),
+            os.path.join(share_lib_dir, "vpi", "VpiCbHdl.cpp"),
+        ],
+        extra_link_args=["-Wl,-rpath,$ORIGIN"],
+    )
+    return _build_lib(libvpi, dist, build_dir)
+
+
+def _build_vhpi_lib(
+    build_dir,
+    include_dir,
+    share_lib_dir,
+    dist,
+    sim_define,
+    extra_lib=[],
+    extra_lib_dir=[],
+):
+    libcocotbvhpi = Extension(
+        "libcocotbvhpi",
+        include_dirs=[include_dir],
+        define_macros=[("VHPI_CHECKING", 1)] + [(sim_define, "")],
+        libraries=["gpi", "gpilog", "stdc++"] + extra_lib,
+        library_dirs=[build_dir] + extra_lib_dir,
+        sources=[
+            os.path.join(share_lib_dir, "vhpi", "VhpiImpl.cpp"),
+            os.path.join(share_lib_dir, "vhpi", "VhpiCbHdl.cpp"),
+        ],
+        extra_link_args=["-Wl,-rpath,$ORIGIN"],
+    )
+
+    return _build_lib(libcocotbvhpi, dist, build_dir)
+
+
+def build(build_dir, debug=False):
+
+    logger = logging.getLogger(__name__)
+
+    if debug:
+        distutils.log.set_verbosity(distutils.log.DEBUG)
+    else:
+        distutils.log.set_verbosity(0)  # Disable logging compilation commands
+
+    cfg_vars = distutils.sysconfig.get_config_vars()
+    for key, value in cfg_vars.items():
+        if type(value) == str:
+            cfg_vars[key] = value.replace("-Wstrict-prototypes", "")
+
+    if sys.platform == "darwin":
+        cfg_vars["LDSHARED"] = cfg_vars["LDSHARED"].replace("-bundle", "-dynamiclib")
+
+    share_dir = os.path.join(os.path.dirname(__file__), "share")
+    share_lib_dir = os.path.join(share_dir, "lib")
+    build_dir = os.path.abspath(build_dir)
+    include_dir = os.path.join(share_dir, "include")
+
+    dist = Distribution()
+    dist.parse_config_files()
+
+    #
+    #  Icarus Verilog
+    #
+    logger.warning("Compiling interface libraries for Icarus Verilog ...")
+    icarus_build_dir = os.path.join(build_dir, "icarus")
+    icarus_compile = True
+    icarus_extra_lib = []
+    icarus_extra_lib_path = []
+    if os.name == "nt":
+        iverilog_path = find_executable("iverilog")
+        if iverilog_path is None:
+            logger.warning(
+                "Icarus Verilog executable not found. No VPI interface will be available."
+            )
+            icarus_compile = False
+        else:
+            icarus_path = os.path.dirname(os.path.dirname(iverilog_path))
+            icarus_extra_lib = ["vpi"]
+            icarus_extra_lib_path = [os.path.join(icarus_path, "lib")]
+
+    if icarus_compile:
+        _build_common_libs(icarus_build_dir, include_dir, share_lib_dir, dist)
```

Comment:
Why do we rebuild the common libraries anew for each simulator? It might be worth a todo about this.

Reply:
see: https://github.com/cocotb/cocotb/pull/1306/files#diff-67c744c69df11e3f27fa958b42c1d390R84-R86

```python
# We need to build common libraries for each simulator to avoid issues with loading
# libraries from different directories in multi os configuration without setting
# globally LD_LIBRARY_PATH or similar.
```